### PR TITLE
Use dropdown without auto completion for simple lists in edit hosting preferences form

### LINKED ISSUE #1268: Use select component from MUI instead of autocomplete for fields with limited options
For the following dropdown fields on the edit hosting preferences form, having autocomplete doesn't make much sense.

Accept smoking
Parking details
Sleeping arangement

Suggestion is to use a Select component (MUI) instead of Autocomplete component.

Note: currently tests and dev-tools are throwing warnings like:

```
console.warn
      Material-UI: The value provided to Autocomplete is invalid.
      None of the options match with `1`.
      You can use the `getOptionSelected` prop to customize the equality test.
      at useAutocomplete (node_modules/@material-ui/lab/useAutocomplete/useAutocomplete.js:267:17)
      at Autocomplete (node_modules/@material-ui/lab/Autocomplete/Autocomplete.js:417:56)
      at renderWithHooks (node_modules/react-dom/cjs/react-dom.development.js:14985:18)
      at updateForwardRef (node_modules/react-dom/cjs/react-dom.development.js:17044:20)
      at beginWork (node_modules/react-dom/cjs/react-dom.development.js:19098:16)
      at beginWork$1 (node_modules/react-dom/cjs/react-dom.development.js:23940:14)
      at performUnitOfWork (node_modules/react-dom/cjs/react-dom.development.js:22779:12)
      at workLoopSync (node_modules/react-dom/cjs/react-dom.development.js:22707:5)
 ```
 
 This should be fixed with this issue.

### PR DESCRIPTION
**Pull request by thdk**
Closes #1268 

Sleeping arrangement (or 'private space'), smoking location and parking details fields on the edit hosting preferences form now no longer allow typing text with autocompletion.

These fields now display the valid options in a dropdown (MUI select component)

Users may have 0 (unspecified) or 1 (unknown) as value for these fields. 
Before this issue, this would log warnings in the console because they were not listed as an option for the autocomplete control.

I have now added the unknown (ask me) option to the dropdown for each of these fields.
If the value of the user was 0 (unspecified) it will fallback in the UI to 1 (unknown) and be saved as 1 (unknown) if the user would submit the form.

The user may now also change a previous saved value to 'ask me'.

If this is not the expected behavior, please let me know.

**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] Added any new components to storybook
- [ ] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes

### CODE REVIEW

File: app/frontend/src/components/Select.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  return (
+    <FormControl
+      variant={variant}
+      id={id}
+      className={classnames(className, classes.formControl)}
+      margin="normal"
+    >
+      <InputLabel htmlFor={id}>{label}</InputLabel>
+      <MuiSelect
+        native
```
Comment (darrenvong) [inline app/frontend/src/components/Select.tsx:47]: I wonder if it's better to not use the `native` option so we get a consistent look and feel between operating systems and also against our other components

**Review by darrenvong — COMMENTED** [review]
LGTM! Maybe the native point I raised should be relayed to Slack to ask for some input from the design team...

File: app/frontend/src/features/profile/edit/EditHostingPreference.tsx [hunk 2]
```diff
@@ -53,7 +54,7 @@ import {
   SleepingArrangement,
   SmokingLocation,
 } from "pb/api_pb";
-import { useState } from "react";
+import React, { useState } from "react";
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/edit/EditHostingPreference.tsx]: revert

File: app/frontend/src/features/profile/edit/EditHostingPreference.tsx [hunk 3]
```diff
@@ -95,6 +96,7 @@ const useStyles = makeStyles((theme) => ({
     paddingTop: theme.spacing(1),
   },
   field: {
+    display: "block",
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/edit/EditHostingPreference.tsx]: Is this necessary? It is also in your new `<Select>` component

File: app/frontend/src/features/profile/edit/EditHostingPreference.tsx [hunk 4]
```diff
@@ -250,31 +252,25 @@ export default function HostingPreferenceForm() {
...
-                forcePopupIcon
-                freeSolo={false}
-                getOptionLabel={(option) => smokingLocationLabels[option]}
-                multiple={false}
-                options={[
+              <Select
+                defaultValue={user.smokingAllowed || SmokingLocation.SMOKING_LOCATION_UNKNOWN}
+                onChange={onChange}
+                label={ACCEPT_SMOKING}
+                name="smokingAllowed"
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/edit/EditHostingPreference.tsx]:
```suggestion
```

Name is set in the controller

File: app/frontend/src/features/profile/edit/EditHostingPreference.tsx [hunk 5]
```diff
@@ -250,31 +252,25 @@ export default function HostingPreferenceForm() {
...
-                    label={ACCEPT_SMOKING}
-                    name="smokingAllowed"
-                    className={classes.field}
-                  />
-                )}
-              />
+                ].map((o) => (
+                  <option value={o} key={o}>
+                    {smokingLocationLabels[o]}
+                  </option>
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/edit/EditHostingPreference.tsx]: Instead of doing this every time, what do you think of giving `<Select>` an `options: {label: string, value: T}[]` prop and doing the map in the component?

File: app/frontend/src/features/profile/edit/EditHostingPreference.tsx [hunk 6]
```diff
@@ -250,31 +252,25 @@ export default function HostingPreferenceForm() {
...
-              <Autocomplete
-                disableClearable={false}
-                defaultValue={user.smokingAllowed}
-                forcePopupIcon
-                freeSolo={false}
-                getOptionLabel={(option) => smokingLocationLabels[option]}
-                multiple={false}
-                options={[
+              <Select
+                defaultValue={user.smokingAllowed || SmokingLocation.SMOKING_LOCATION_UNKNOWN}
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/edit/EditHostingPreference.tsx]: Seems like code smell to have `defaultValue` on both the controller and the rendered component. Maybe the most correct way would be to set `value` on the component from `render={({ onChange, value }) => (` ? Not totally sure

**Review by lucaslcode — CHANGES_REQUESTED** [review]

File: app/frontend/src/features/profile/edit/EditHostingPreference.tsx (hunk 6 above)
Comment (thdk) [reply app/frontend/src/features/profile/edit/EditHostingPreference.tsx]: yes :( seems it didn't work indeed. I've added tests now to check if the correct value from the database is set in the form.

File: app/frontend/src/features/profile/edit/EditHostingPreference.tsx (hunk 5 above)
Comment (thdk) [reply app/frontend/src/features/profile/edit/EditHostingPreference.tsx]: will do :)

File: app/frontend/src/components/Select.tsx (hunk 1 above)
Comment (thdk) [reply app/frontend/src/components/Select.tsx:47]: "As the user experience can be improved on mobile using the native select of the platform, we allow such pattern."
https://material-ui.com/components/selects/#native-select

File: app/frontend/src/features/profile/edit/EditHostingPreference.tsx [hunk 7]
```diff
@@ -249,28 +248,55 @@ export default function HostingPreferenceForm() {
...
+                        SmokingLocation.SMOKING_LOCATION_OUTSIDE
+                      ],
+                  },
+                  {
+                    value: SmokingLocation.SMOKING_LOCATION_NO,
+                    label:
+                      smokingLocationLabels[
+                        SmokingLocation.SMOKING_LOCATION_NO
+                      ],
+                  },
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/edit/EditHostingPreference.tsx]: Haha this is worse...
What if you do `Object.entries(smokingLocationLabels).map(([key, value]) => ({label: value, value: +key}))`?
However that may be  less type safe and one could argue it's more opaque.

You decide if you want to leave it like you have done now, how it was before, or that one-liner.

**Review by lucaslcode — CHANGES_REQUESTED** [review]
One comment - if you want to leave as is, re-request and I'll accept

File: app/frontend/src/features/profile/edit/EditHostingPreference.tsx (hunk 7 above)
Comment (thdk) [reply app/frontend/src/features/profile/edit/EditHostingPreference.tsx]: I agree it looks worse, but it is more 'solid' because at least you don't have to write html tags (but still can using the children prop of select). I've tried it with object.entries but then we are less flexibile in the sort order of options and you can't have entries in the labels which you don't want to display in the dropdown. 
Eg. the labels contains both a label for unknown and unspecified while now I only show unknown in the dropdown.

Another option is to pass the select component an array of values and a dictionary (or a getLabel function) from which to retrieve the label for each value.

Comment (lucaslcode) [reply app/frontend/src/features/profile/edit/EditHostingPreference.tsx]: I think that's a good idea - our usual use case is the protobuf enums, and in other cases, it probably won't be much more verbose than any other way

File: app/frontend/src/components/Select.tsx [hunk 8]
```diff
@@ -20,19 +20,41 @@ const useStyles = makeStyles((theme) => ({
...
   id: string;
-  options?: { label: string; value: string | number }[];
+  options?: Extract<keyof T, string | number>[];
+  value?: T extends undefined ? string | number : keyof T;
+  data?: T;
 }) {
   const classes = useStyles();
+
+  const SelectOptions = () => {
+    if (!options || !data) {
```
Comment (lucaslcode) [inline app/frontend/src/components/Select.tsx]: If this is the case, those props shouldn't be optional and you won't need this

File: app/frontend/src/components/Select.tsx [hunk 9]
```diff
@@ -20,19 +20,41 @@ const useStyles = makeStyles((theme) => ({
...
   label,
   variant = "outlined",
   options,
   ...otherProps
 }: SelectProps & {
   id: string;
-  options?: { label: string; value: string | number }[];
+  options?: Extract<keyof T, string | number>[];
+  value?: T extends undefined ? string | number : keyof T;
+  data?: T;
```
Comment (lucaslcode) [inline app/frontend/src/components/Select.tsx]: If you make the type `T extends Record<string | number, string>`, maybe you don't need `options` at all?

And what do you think about making the name `labelledOptions` or `optionsAndLabels` or something?

**Review by lucaslcode — CHANGES_REQUESTED** [review]

File: app/frontend/src/components/Select.tsx (hunk 8 above)
Comment (thdk) [reply app/frontend/src/components/Select.tsx]: You can still provide options using the `children `prop. In that case `options `and `data `can be undefined.

File: app/frontend/src/components/Select.tsx (hunk 9 above)
Comment (thdk) [reply app/frontend/src/components/Select.tsx]: The reason I need both `options` and `data` is because the labels object contains more items than I want to display in the dropdown. I don't see how that would be different when using Record (which seems a better option here indeed) than an index signature.

Comment (lucaslcode) [reply app/frontend/src/components/Select.tsx]: Oh, I see. Can you think of a better name for `data`? I wouldn't know the difference between it and `options` without looking at a previous usage or the source.

File: app/frontend/src/components/Select.tsx (hunk 8 above)
Comment (lucaslcode) [reply app/frontend/src/components/Select.tsx]: Is that necessary?

File: app/frontend/src/components/TextField.tsx [hunk 10]
```diff
@@ -37,7 +36,6 @@ export default function TextField({
     <MuiTextField
       {...otherProps}
       variant={variant}
-      id={id}
```
Comment (darrenvong) [inline app/frontend/src/components/TextField.tsx:40]: Why was this removed? We need this previously so [MUI correctly links label up with the right ID](https://material-ui.com/components/text-fields/#accessibility). Removing it will break this

File: app/frontend/src/components/Select.tsx [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  ...otherProps
+}: SelectProps & {
+  id: string;
+  options?: Extract<keyof T, string | number>[];
+  value?: T extends undefined ? string | number : keyof T;
+  data?: T;
+}) {
+  const classes = useStyles();
+
+  const SelectOptions = () => {
```
Comment (darrenvong) [inline app/frontend/src/components/Select.tsx]: Can we lift this out of this component? Creating a component inside a component is an anti-pattern as it creates a new function reference each time so React won't know it's the same component and so screws up its reconciliation and ends up recreating the component tree and replacing more things in the DOM than necessary

File: app/frontend/src/components/Select.tsx (hunk 8 above)
Comment (thdk) [reply app/frontend/src/components/Select.tsx]: no, but it's cool :) (I'll remove it)

File: app/frontend/src/components/TextField.tsx (hunk 10 above)
Comment (thdk) [reply app/frontend/src/components/TextField.tsx:40]: I've also removed  `id` from the destructured props. So `otherProps` will still have the `id` prop. The spread operator used on line 38 will nicely pass all these props, including the `id` prop, to the MuiTextField component.

So this change is just a 'code style' issue.

File: app/frontend/src/components/Select.tsx (hunk 11 above)
Comment (thdk) [reply app/frontend/src/components/Select.tsx]: I can lift this out of the component. But I can also use useMemo for this. Any preference?

File: app/frontend/src/components/Select.tsx (hunk 9 above)
Comment (thdk) [reply app/frontend/src/components/Select.tsx]: what about 'optionLabelMap' ?

Comment (lucaslcode) [reply app/frontend/src/components/Select.tsx]: Sounds good

File: app/frontend/src/components/Select.tsx (hunk 11 above)
Comment (lucaslcode) [reply app/frontend/src/components/Select.tsx]: Moving it out is better, but in this case you can just inline the options instead of having another component

**Review by lucaslcode — APPROVED** [review]
Great! Pretty sure all Darren's comments are addressed too so I'll approve :)
